File: breadth_first_search/0127_word_ladder/0127_word_ladder.py

class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        from string import ascii_lowercase
        distance, cur, visited, lookup = 0, [beginWord], set([beginWord]), set(wordList)

        while cur:
            next_queue = []

            for word in cur:
                if word == endWord:
                    return distance + 1
                for i in xrange(len(word)):
                    for j in ascii_lowercase:
                        candidate = word[:i] + j + word[i+1:]
                        if candidate not in visited and candidate in lookup:
                            next_queue.append(candidate)
                            visited.add(candidate)
            distance += 1
            cur = next_queue
        return 0 


beginWord = "hit"
endWord = "cog"
wordList = ["hot","dot","dog","lot","log","cog"]

res = Solution().ladderLength(beginWord, endWord, wordList)
print(res)

File: breadth_first_search/0127_word_ladder/0127_word_ladder.go
package main

import (
	"fmt"
)

func ladderLength(beginWord, endWord string, words []string) int {
	dict := make(map[string]bool, len(words))
	for i := 0; i < len(words); i++ {
		dict[words[i]] = true
	}

	delete(dict, beginWord)

	queue := make([]string, 0, len(words))

	var trans func(string) bool
	trans = func(word string) bool {
		bytes := []byte(word)
		for i := 0; i < len(bytes); i++ {
			diffLetter := bytes[i]
			for j := 0; j < 26; j++ {
				b := 'a' + byte(j)
				if b == diffLetter {
					continue
				}

				bytes[i] = b

				if dict[string(bytes)] {
					if string(bytes) == endWord {
						return true
					}

					queue = append(queue, string(bytes))
					delete(dict, string(bytes))
				}
			}

			bytes[i] = diffLetter
		}

		return false
	}

	queue = append(queue, beginWord)
	dist := 1

	for len(queue) > 0 {
		qLen := len(queue)

		for i := 0; i < qLen; i++ {
			word := queue[0]
			queue = queue[1:]

			if trans(word) {
				return dist + 1
			}
		}

		dist++
	}

	return 0
}

func main() {
	beginWord := "hit"
	endWord := "cog"
	words := []string{"hot", "dot", "dog", "lot", "log", "cog"}
	res := ladderLength(beginWord, endWord, words)
	fmt.Println(res)
}
